src/types.jl

```diff
@@ -14,82 +14,152 @@ end
 
 include("interaction/iodevices.jl")
 
+
 """
-This struct provides accessible `Observable`s to monitor the events
+This struct provides accessible `PriorityObservable`s to monitor the events
 associated with a Scene.
 
+Functions that act on a `PriorityObservable` must return true if the function 
+consumes an event and false if it does not. When an event is consumed it does 
+not trigger other observer functions. The order in which functions are exectued
+can be controlled via the `priority` keyword (default 0) in `on`.
+
+Example:
+```
+on(events(scene).mousebutton, priority = Int8(20)) do event
+    if is_correct_event
+        do_something()
+        return true
+    end
+    return false
+end
+```
+
 ## Fields
 $(TYPEDFIELDS)
 """
 struct Events
     """
     The area of the window in pixels, as a `Rect2D`.
     """
-    window_area::Node{IRect2D}
+    window_area::PriorityObservable{IRect2D}
     """
     The DPI resolution of the window, as a `Float64`.
     """
-    window_dpi::Node{Float64}
+    window_dpi::PriorityObservable{Float64}
     """
     The state of the window (open => true, closed => false).
     """
-    window_open::Node{Bool}
+    window_open::PriorityObservable{Bool}
 
     """
-    The pressed mouse buttons.
-    Updates when a mouse button is pressed.
+    Most recently triggered `MouseButtonEvent`. Contains the relevant 
+    `event.button` and `event.action` (press/release)
 
     See also [`ispressed`](@ref).
     """
-    mousebuttons::Node{Set{Mouse.Button}}
+    mousebutton::PriorityObservable{MouseButtonEvent}
     """
-    The position of the mouse as a `NTuple{2, Float64}`.
-    Updates whenever the mouse moves.
+    A Set of all currently pressed mousebuttons.
     """
-    mouseposition::Node{NTuple{2, Float64}}
+    mousebuttonstate::Node{Set{Mouse.Button}}
```

Comment:
`mousebuttonstate` and `keyboardstate` are basically what `mousebuttons` and `keyboardbuttons` are on master, though `keyboardstate` doesn't trigger on repeats. They're still useful to have (e.g. for `ispressed`) but I'm not sure in what form.

For compat it would be nice to have them as `mousebuttons` and `keyboardbuttons` nodes, doing the same thing they used to.

Conceptually in the new design they shouldn't be nodes - you should react to `keyboardbutton` and `mousebutton` instead. They should also be easy to differentiate from those names.

So what should we do here? We could also do both and have the old `mousebuttons` and `keyboardbuttons` around for deprecation purposes while implementing buttonstates as sets.

Reply:
I changed states to be just `Set`s and added a `getproperty` that generates nodes for `mousebuttons` and `keyboardbuttons` which should behave like on master.